fv3gfs/util/halo_data_transformer.py

```diff
@@ -0,0 +1,916 @@
+    key,
+    shape,
+    slices: Tuple[slice],
+    dims,
+    strides,
+    itemsize: int,
+    rotate: bool,
+    rotation: int,
+) -> "cp.ndarray":
+    """Build an array of indexing from a slice & memory description to build an indexation into the "flatten" memory.
+    
+    Go from a memory layout (strides, itemsize, shape) and slices into it to a
+    single array of indices. We leverage numpy iterator and calculate from the multi_index
+    using memory layout the index into the original memory buffer.
+    """
+
+    # Have to go down to numpy to leverage indices calculation
+    arr_indices = np.empty(shape, dtype=np.int32, order="C")[slices]
+
+    # Get offset from first index
+    offset_dims = []
+    for s in slices:
+        offset_dims.append(s.start)
+    offset_to_slice = sum(np.array(offset_dims) * strides) // itemsize
+
+    # Flatten the index into an indices array
+    with np.nditer(
+        arr_indices, flags=["multi_index"], op_flags=["writeonly"], order="K",
+    ) as it:
+        for array_value in it:
+            offset = sum(np.array(it.multi_index) * strides) // itemsize
+            array_value[...] = offset_to_slice + offset
+
+    if rotate:
+        # sending data across the boundary will rotate the data
+        # n_clockwise_rotations times, due to the difference in axis orientation.
+        # Thus we rotate that number of times counterclockwise before sending,
+        # to get the right final orientation. We apply those rotations to the
+        # indices here to prepare for a straightforward copy in cu kernel
+        arr_indices = rotate_scalar_data(arr_indices, dims, cp, -rotation)
+    return cp.asarray(arr_indices.flatten(order="C"))
+
+
+# ------------------------------------------------------------------------
+# HaloDataTransformer helpers
+
+
+def _slices_size(slices: Tuple[slice, ...]) -> int:
+    """Compute linear size from slices."""
+    length = 1
+    for s in slices:
+        assert s.step is None
+        length *= abs(s.start - s.stop)
+    return length
+
+
+@dataclass
+class HaloExchangeData:
+    """Memory description of the data exchanged.
+    
+    The data stored here target a single exchange, with an optional
+    rotation to give prior to pack. Slices are tupled following the
+    convention of one slice per dimension
+
+    Args:
+        specification: memory layout of the data
+        pack_slices: indexing to pack, one slice per dimension
+        pack_clockwise_rotation:  number of 90-degree rotations to perform
+            before packing
+        unpack_slices: indexing to unpack, one slice per dimension
+    """
+
+    specification: HaloUpdateSpec
+    pack_slices: Tuple[slice, ...]
+    pack_clockwise_rotation: int
+    unpack_slices: Tuple[slice, ...]
+
+    def __post_init__(self):
+        self._id = uuid1()
+        self._pack_buffer_size = _slices_size(self.pack_slices)
+        self._unpack_buffer_size = _slices_size(self.unpack_slices)
```

Comment:
These are accessed by other classes and should be public variables.

Reply:
Good catch